ksqldb-engine/src/main/java/io/confluent/ksql/query/KafkaStreamsQueryValidator.java

```diff
@@ -69,10 +76,43 @@ private void validateCacheBytesUsage(
       return;
     }
     final long configured = getCacheMaxBytesBuffering(config);
-    final long usedByRunning = running.stream()
-        .mapToLong(r -> new StreamsConfig(r.getStreamsProperties())
-                .getLong(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG))
-        .sum();
+    long usedByRunning;
+    if (!config.getConfig(true).getBoolean(KsqlConfig.KSQL_SHARED_RUNTIME_ENABLED)) {
+      usedByRunning = running.stream()
+          .mapToLong(r -> new StreamsConfig(r.getStreamsProperties())
+              .getLong(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG))
+          .sum();
+    } else {
+      usedByRunning = running.stream()
+          .filter(t -> !(t instanceof BinPackedPersistentQueryMetadataImpl))
+          .mapToLong(r -> new StreamsConfig(r.getStreamsProperties())
+              .getLong(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG))
+          .sum();
+      final Set<String> runtimes = new HashSet<>();
+      long cacheSizeBytesPerRuntime = -1L;
+      for (final QueryMetadata queryMetadata : running) {
+        if (queryMetadata instanceof BinPackedPersistentQueryMetadataImpl) {
+          if (cacheSizeBytesPerRuntime == -1L) {
+            cacheSizeBytesPerRuntime = (long) queryMetadata.getStreamsProperties()
+                .get(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG);
+          }
+          if (!runtimes.contains(queryMetadata.getQueryApplicationId())) {
+            runtimes.add(queryMetadata.getQueryApplicationId());
+            usedByRunning += (long) queryMetadata.getStreamsProperties()
+                .get(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG);
+          } else {
```

Comment:
nit: technically we should also check/verify that the cache size is the same across all runtimes, not just within the runtime. Since `ALTER SYSTEM` applies to all runtimes* there should never be a discrepancy in the server level cache size config

*one exception to this would be if there are gen 1 queries with different cache sizes running on an upgraded gen 2 cluster. Maybe we can distinguish these by the application id since only gen 1 queries will contain the query id in the app id? Or is there a better way to differentiate these -- eg do gen 1 queries still use `BinPackedPersistentQueryMetadata`  or are they just regular `PersistentQueryMetadata`?